Add opsgenie plugin

EdwardDowling opened the pull request:
Initial opsgenie plugin
Part of [437](https://github.com/gravitational/teleport-plugins/issues/437)
Changes to /e will be in a separate PR

EdwardDowling committed:
Move opsgenie client code to integrations/access

EdwardDowling committed:
Add initial opsgenie bot skeleton and stubs

EdwardDowling committed:
Add opsgenie check and set default to pluginv1

EdwardDowling committed:
Fix check and set defaults for opsgenie plugin config

EdwardDowling committed:
Fill out bot stubs for opsgenie bot

EdwardDowling committed:
Add loadOpsgenieConfig to integrations

EdwardDowling committed:
Swap to using bearer token for opsgenie plugin

EdwardDowling committed:
Fix formating in opsgenie bot

EdwardDowling committed:
Reorder imports to fix lint errors

tigrato commented on integrations/access/opsgenie/app.go:
```diff
@@ -0,0 +1,31 @@
+/*
+Copyright 2022 Gravitational, Inc.
```
```suggestion
Copyright 2023 Gravitational, Inc.
```

tigrato commented on integrations/access/opsgenie/bot.go:
```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2022 Gravitational, Inc.
```
```suggestion
Copyright 2023 Gravitational, Inc.
```

tigrato commented on integrations/access/opsgenie/bot.go:
```diff
@@ -0,0 +1,106 @@
+}
+
+// CheckHealth checks if the bot can connect to its messaging service
+func (b *Bot) CheckHealth(ctx context.Context) error {
```
Is it supposed to skip the test?

justinas replied:
➕ this raises eyebrows - does not explain why it is not supported for opsgenie. Is there no "ping" endpoint we could call that would just check that we're authenticated correctly / API is up?

EdwardDowling committed:
Add config field to plugins to allow for different recipient logic

Allows bots that use schedules as recipients to determine their own
logic for defaulting etc

EdwardDowling committed:
Add notimplemented err for check health for opsgenie bots

justinas commented on api/types/plugin.go:
```diff
@@ -111,6 +113,21 @@ func (p *PluginV1) CheckAndSetDefaults() error {
+		if bearer == nil {
+			return trace.BadParameter("opsgenie plugin must be used with the bearer token credential type")
+		}
+		if (bearer.Token == "") == (bearer.TokenFile == "") {
```
Both this and the Okta plugin add the same validation rule. Might be smart to factor this out to some `func (*PluginBearerTokenCredentials) CheckAndSetDefaults() error`

https://github.com/gravitational/teleport/pull/26349/files#diff-c05dcfe52aba6413a7c9443e99c3204098909bc87148471197e941693cb09614R132-R134
@mdwn

justinas commented on integrations/access/common/config.go:
```diff
@@ -35,12 +35,15 @@ type PluginConfiguration interface {
 	Teleport   lib.TeleportConfig
 	Recipients RawRecipientsMap `toml:"role_to_recipients"`
 	Log        logger.Config
+	// RecipientsAreSchedules is used to mark wether the recipients of alerts are on-call schedules.
```
Not sure I like introducing the seemingly opsgenie (and maybe pagerduty) specific concept of schedules into the `BaseConfig` / `BaseApp`. What about `bool UsersAsRecipients` or similar, that would indicate whether individual assigned reviewers on the access requests should be considered valid recipients? Could be `true` for Slack/Discord/similar, false for others.

EdwardDowling replied:
That sounds a lot clearer.

justinas commented on integrations/access/opsgenie/bot.go:
```diff
@@ -0,0 +1,110 @@
+	for _, recipient := range b.client.DefaultSchedules {
+		schedules = append(schedules, recipient)
+	}
+	opsgeneieReqData := RequestData{
```
Typo in variable name

EdwardDowling committed:
Add check health to opsgenie client and bot

EdwardDowling committed:
Rename recipientsAreSchedules to usersAsRecipients

EdwardDowling committed:
Add pluginBearertokencredentials checkAndSetdefaults

EdwardDowling committed:
Add resolveAnnotations to reqdata and use schedules from that

r0mant commented on api/types/plugin.go:
```diff
@@ -257,6 +282,14 @@ func (s *PluginSlackAccessSettings) CheckAndSetDefaults() error {
+// CheckAndSetDefaults validates and set the default values
+func (s *PluginOpsgenieAccessSettings) CheckAndSetDefaults() error {
+	if s.Addr == "" {
+		return trace.BadParameter("opsgenie addr must be set")
```
What addr is this? API endpoint? Let's say so explicitly in the error message.

r0mant commented on api/types/plugin.go:
```diff
@@ -257,6 +282,14 @@ func (s *PluginSlackAccessSettings) CheckAndSetDefaults() error {
 
+// CheckAndSetDefaults validates and set the default values
+func (s *PluginOpsgenieAccessSettings) CheckAndSetDefaults() error {
+	if s.Addr == "" {
```
If this is Opsgenie API endpoint, can this field be renamed to `APIEndpoint`? `Addr` is too generic.

r0mant commented on integrations/access/common/config.go:
```diff
@@ -35,12 +35,16 @@ type PluginConfiguration interface {
 	Log        logger.Config
+	// UsersAsRecipients is set for bots where the individual assigned
+	// reviewers on the access requests should be considered valid recipients.
+	UsersAsRecipients bool
```
I'm not sure I understand this tbh. Could you give an example of why this is different from existing plugins?

EdwardDowling replied:
For the slack plugin the the recipients of the access request  get passed to get message recipients which does not make sense for the opsgenie plugin. This is to let onPendingRequest know not to call that and instead let the plugin figure it out itself.

r0mant replied:
But why do we need this additional flag on the base config? Can we just override `GetRecipients()` on the Opsgenie config method and have it return appropriate recipients?

r0mant commented on integrations/access/opsgenie/config.go:
```diff
@@ -0,0 +1,125 @@
+
+// NewBot initializes the new Opsgenie message generator (OpsgenieBot)
+func (c *Config) NewBot(clusterName, webProxyAddr string) (common.MessagingBot, error) {
+
```
```suggestion
```

EdwardDowling committed:
Update integrations/access/opsgenie/bot.go

Co-authored-by: Roman Tkachenko <roman@goteleport.com>

EdwardDowling committed:
Prevent default schedules being processed if annotations are set

EdwardDowling committed:
Remove loadOpsgenieConfig

EdwardDowling committed:
Rename opsgneie addr field

EdwardDowling committed:
Add more verbose error messages to opsgenie client

r0mant commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -5066,8 +5066,8 @@ message PluginSlackAccessSettings {
 
-  // Addr is the address of Opsgenie API.
-  string addr = 1;
+  // API_endpoint is the address of Opsgenie API.
+  string API_endpoint = 1;
```
```suggestion
  // APIEndpoint is the address of Opsgenie API.
  string api_endpoint = 1;
```

r0mant commented on integrations/access/opsgenie/bot.go:
```diff
@@ -0,0 +1,113 @@
+func (b *Bot) Broadcast(ctx context.Context, _ []common.Recipient, reqID string, reqData pd.AccessRequestData) (data common.SentMessages, err error) {
+	schedules := []string{}
+
+	if _, ok := reqData.ResolveAnnotations[ReqAnnotationRespondersKey]; ok {
+		schedules = reqData.ResolveAnnotations[ReqAnnotationRespondersKey]
```
```suggestion
	if recipients, ok := reqData.ResolveAnnotations[ReqAnnotationRespondersKey]; ok {
		schedules = append(schedules, recipients...)
```

r0mant commented on integrations/access/opsgenie/client.go:
```diff
@@ -220,11 +223,29 @@ func (og Client) GetOnCall(ctx context.Context, scheduleName string) (Responders
+// CheckHealth pings opsgenie.
+func (og Client) CheckHealth(ctx context.Context) error {
+	// The heartbeat pings will respond even if the heartbeat does not exist.
+	resp, err := og.client.NewRequest().
+		SetContext(ctx).
+		SetPathParams(map[string]string{"heartbeat": heartbeatName}).
+		Get("heartbeats/{heatbeat}/ping")
```
I'm looking at Opsgenie API and it seems like you're supposed to create the heartbeat first before you can use it?

https://docs.opsgenie.com/docs/heartbeat-api#add-heartbeat-request

If so, I don't think we can use it as a health check as we would have to ask users to set up the heartbeat first which we probably don't want to require.

Is there another API we can use to check? Maybe just try and retrieve the schedules, or some other read-only call the API key we ask for allows to use? Basically, something the plugin will have to do anyway when creating an incident.

EdwardDowling replied:
From what i understood from "Since Opsgenie process heartbeat ping requests asynchronously, it does not check if the heartbeat exists or not before responding.
Please note that receiving a PONG response does not necessarily mean that the heartbeat exists." it should be fine to rely on this responding even if the heartbeat does not exist.

We could also just create the heartbeat when the first healthcheck is made as an alternative?

EdwardDowling committed:
Update api/proto/teleport/legacy/types/types.proto

Co-authored-by: Roman Tkachenko <roman@goteleport.com>

EdwardDowling committed:
Update integrations/access/opsgenie/bot.go

Co-authored-by: Roman Tkachenko <roman@goteleport.com>

EdwardDowling committed:
Fix error message for recipients field in opsgenie bot

EdwardDowling committed:
Change apiEndpoint field name

EdwardDowling committed:
Remove check for unused field from opsgenie config

EdwardDowling commented:
> But why do we need this additional flag on the base config? Can we just override GetRecipients() on the Opsgenie config method and have it return appropriate recipients?

This was done to avoid going through getMessageRecipients entirely as that uses conf.Recipients which is not used by the opsngenie bot. Also it would require the suggested reviewers be valid emails which would not make sense for schedules.

EdwardDowling commented on integrations/access/opsgenie/bot.go:
```diff
@@ -0,0 +1,113 @@
+)
+
+const (
+	ScheduleRecipientKind = "schedule"
```
It is now used in FetchRecipients

EdwardDowling committed:
Remove usersAsRecipients flag

espadolini commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -5102,8 +5102,8 @@ message PluginSlackAccessSettings {
-  // Addr is the address of Opsgenie API.
-  string addr = 1;
+  // APIEndpoint is the address of Opsgenie API.
+  string api_endpoint = 1;
```
This breaks JSON marshaling and unmarshaling of `PluginOpsgenieAccessSettings`, both in the `encoding/json` sense and in the `protojson` sense. If the break was deliberate then remove the addr field with number 1, add `reserved 1; reserved "addr";` and add `api_endpoint` with number 5. to the message.

r0mant replied:
@espadolini The `PluginOpsgenieAccessSettings` message is present in v13 but it is not used by anything yet. Is renaming this field still an issue in this case?

espadolini replied:
It's still very bad practice in general, and why add to the mental load of anyone who is going to look at this change? Field numbers and field names are free.

r0mant commented on integrations/access/opsgenie/client.go:
```diff
@@ -59,6 +61,8 @@ type Client struct {
 
 	client *resty.Client
+
+	heartbeatCreated bool
```
This seems to be unused.
```suggestion
```

r0mant commented on integrations/access/common/recipient.go:
```diff
@@ -23,6 +23,16 @@ import (
+
+const (
+	// ReqAnnotationRespondersKey is the request annotation key at which alternate responders are stored.
+	ReqAnnotationRespondersKey = "teleport.dev/responder"
```
What is a "responder"? Is it a schedule name basically?

EdwardDowling replied:
swapped to schedules

r0mant reviewed (approved):
lgtm once the .proto issue @espadolini pointed out is addressed (if it's an issue)

EdwardDowling committed:
Reserve addr and change change field number of api_endpoint

EdwardDowling committed:
Update integrations/access/opsgenie/client.go

Co-authored-by: Roman Tkachenko <roman@goteleport.com>

EdwardDowling committed:
Rename reqAnnotationresponderskey to reqannotationscheduleskey

EdwardDowling committed:
Remove unused check and set defaults

EdwardDowling commented:
@tigrato @justinas could you have another look when you have a chance?

tigrato commented on integrations/access/common/recipient.go:
```diff
@@ -23,6 +23,16 @@ import (
+
+const (
+	// ReqAnnotationSchedulesKey is the request annotation key at which schedules are stored for access plugins.
+	ReqAnnotationSchedulesKey = "teleport.dev/schedules"
```
Move the label to package `types` where we have all the labels defined. It's easier to reason about the labels we use.

Use the `types.TeleportNamespace` for namespace

tigrato commented on integrations/access/opsgenie/plugindata.go:
```diff
@@ -60,7 +57,7 @@ type RequestData struct {
 	ReviewsCount       int
 	Resolution         Resolution
-	RequestAnnotations map[string][]string
+	ResolveAnnotations map[string][]string
```
type.Labels is an alias for this

tigrato reviewed (approved):
Just two minor comments

EdwardDowling committed:
Rename REqAnnotationScheduleskey

EdwardDowling committed:
Use types.Labels alias where possible

EdwardDowling committed:
Merge branch 'master' into edwarddowling/hosted-opsgenie

EdwardDowling committed:
Simplify loop in opsgenie bot to satisfy linter
